tree/ntuple/v7/src/RField.cxx

```diff
@@ -3416,6 +3422,106 @@ ROOT::Experimental::RUniquePtrField::SplitValue(const RValue &value) const
+   : RNullableField(fieldName, typeName, std::move(itemField)), fItemDeleter(GetDeleterOf(*fSubFields[0]))
+{
+   if (fSubFields[0]->GetTraits() & kTraitTriviallyDestructible)
+      fTraits |= kTraitTriviallyDestructible;
+}
+
+std::pair<void *, bool *> ROOT::Experimental::ROptionalField::GetValueAndEngagementPtrs(void *optionalPtr) const
+{
+   void *value = optionalPtr;
+   bool *engagement =
+      reinterpret_cast<bool *>(reinterpret_cast<unsigned char *>(optionalPtr) + fSubFields[0]->GetValueSize());
+   return {value, engagement};
+}
+
+std::pair<const void *, const bool *>
+ROOT::Experimental::ROptionalField::GetValueAndEngagementPtrs(const void *optionalPtr) const
+{
+   return GetValueAndEngagementPtrs(const_cast<void *>(optionalPtr));
+}
+
+std::unique_ptr<ROOT::Experimental::RFieldBase>
+ROOT::Experimental::ROptionalField::CloneImpl(std::string_view newName) const
+{
+   auto newItemField = fSubFields[0]->Clone(fSubFields[0]->GetFieldName());
+   return std::make_unique<ROptionalField>(newName, GetTypeName(), std::move(newItemField));
+}
+
+std::size_t ROOT::Experimental::ROptionalField::AppendImpl(const void *from)
+{
+   const auto [valuePtr, engagementPtr] = GetValueAndEngagementPtrs(from);
+   if (*engagementPtr) {
+      return AppendValue(valuePtr);
+   } else {
+      return AppendNull();
+   }
+}
+
+void ROOT::Experimental::ROptionalField::ReadGlobalImpl(NTupleSize_t globalIndex, void *to)
+{
+   auto [valuePtr, engagementPtr] = GetValueAndEngagementPtrs(to);
+   auto itemIndex = GetItemIndex(globalIndex);
+   if (itemIndex.GetIndex() == kInvalidClusterIndex) {
+      *engagementPtr = false;
+   } else {
+      CallReadOn(*fSubFields[0], itemIndex, valuePtr);
+      *engagementPtr = true;
+   }
+}
+
+void ROOT::Experimental::ROptionalField::ConstructValue(void *where) const
+{
+   auto [valuePtr, engagementPtr] = GetValueAndEngagementPtrs(where);
+   CallConstructValueOn(*fSubFields[0], valuePtr);
+   *engagementPtr = false;
+}
+
+void ROOT::Experimental::ROptionalField::ROptionalDeleter::operator()(void *objPtr, bool dtorOnly)
+{
+   fItemDeleter->operator()(objPtr, true /* dtorOnly */);
+   RDeleter::operator()(objPtr, dtorOnly);
+}
+
+std::unique_ptr<ROOT::Experimental::RFieldBase::RDeleter> ROOT::Experimental::ROptionalField::GetDeleter() const
+{
+   return std::make_unique<ROptionalDeleter>(GetDeleterOf(*fSubFields[0]));
+}
+
+std::vector<ROOT::Experimental::RFieldBase::RValue>
+ROOT::Experimental::ROptionalField::SplitValue(const RValue &value) const
+{
+   std::vector<RValue> result;
+   const auto [valuePtr, engagementPtr] = GetValueAndEngagementPtrs(value.GetPtr<void>().get());
+   if (*engagementPtr) {
+      result.emplace_back(fSubFields[0]->BindValue(std::shared_ptr<void>(value.GetPtr<void>(), valuePtr)));
+   }
+   return result;
+}
+
+size_t ROOT::Experimental::ROptionalField::GetValueSize() const
+{
+   const auto alignment = GetAlignment();
+   const auto actualSize = fSubFields[0]->GetValueSize() + 1;
```

Comment:
What's `1`? I assume it is `sizeof(char)` i.e. `sizeof(engagement)`.

Reply:
or is it to make the modulo below work?

Reply:
Clarified in https://github.com/root-project/root/pull/15720